### DISCUSSION TITLE
Of Trees and Forests

### DISCUSSION THREAD

**Discussion by JohnG-cc** [CredComSoc/ProtocolWorkingGroup#6] [Track 2 Ideas, requirements and features]
A tree of nodes is a wonderfully simplifying heuristic for network routing and also for eliminating the need to find credit clearing loops. 

The fact that the Credit Commons Protocol is intended to be an open standard with open source implementations, strongly suggests that there will eventually be more than one tree in existence. 

And *that* implies that, just as any given branch or leaf node in a tree can add sub-branches without seeking anyone's permission or even notifying their trunkwards nodes, any given node can *also have BoT accounts for multiple independent trees* if they choose to have their software support that. 

It seems to me like this is more of a feature than a bug. One could imagine a geographically organised tree, where there is a long and many-node path from Auckland/NorthIsland/NZ/Oceania/creditcommons.org to Brixton/North/UK/Neighbours/Europe/creditcommons.org, and also imagine another tree organised around industry verticals, where AucklandLabs/SoftwareServices/IT&T/communitycode.org and BrixtonStudios/SoftwareServices/IT&T/communitycode.org were right next to each other. 

(wait, I have those paths backwards, don't I? Also, note the requirement for a trunk name now, since it can no longer be assumed)

This would provide redundant routing in case of outage/attack/confiscation, and if done right, would exempt the various trunk nodes from having to process the entire world's traffic through a single system - all while retaining most of the benefits of a tree based path structure.

A first-brush overview of how such a forest could be set up would by by layering a Path Resolution Protocol on top of the Credit Commons Protocol. This would allow the payer to discover the various paths under which the payee can be reached, and ideally to learn the fees and balance-bandwith available over each path. The payer could then select one and make payment using the Credit Commons Protocol.

This might look something like the payee providing the payer with a URL rather than a path, something like www.payeesNode.org/payeeAccountNumber. The payer's node could then query that URL, and get back a list of every single path that that account is known as:

    {
      "account_name": "BrixtonStudios",
      "paths": [
        "BrixtonStudios/Brixton/North/UK/Neighbours/Europe/creditcommons.org",
        "BrixtonStudios/SoftwareServices/IT&T/communitycode.org",
      ]
    }

The payees node can then filter that list for trunks that it is also part of, and issue light-weight queries to it's relevant trunkwards nodes using the PRP protocol, to determine available balance-bandwidth and relative fees. Having selected the optimum node either automatically or by user interaction, it can then send a Credit Commons transaction via the tree it has chosen.

This actually generates a dramatic increase in total chatter between nodes, that still needs to be routed via each trunk - however these do not need balance reservations and timeout, nor do they need to be persisted to the database which makes it MUCH easier for these queries to be handled at high throughput. Also, given a large number of trunks (think as in Mastodon server type of distribution), it's not likely that any one trunk would end up handling the whole world's traffic, since it's unlikely any payee node would be on every trunk, and even then the payer would probably not be either.

A further refinement would be allowing a transaction to jump between trees at connecting nodes, but here we would be getting into something of a routing protocol and this might better be left for a version 3.0, if at all....

@matslats @dilgreen @utunga @tfwoodroofMCS what do you think?

**Comment by matslats** [comment]
Dil and I have discussed this.
In my terminology the a network which allowed multiple trunkward nodes would switch from being a tree to being a mesh, and lose all its elegance. Economically speaking the tree structure means that each node has a single balance of trade with respect to all its peers, two trade balances is twice as hard to manage as one, at least.
It also makes it a bit harder to manage credit limits of peers who might have other debts that you can't see.
All in all I'm very reluctant to go there because it makes it an order of magnitude more complicated OR is better handled as a straight mesh, with no attempt to collectively balance trade. (For me balancing trade is an axiom of justice which neoliberalism has dismissed.)
We could get imaginative about how to do this, but I would start by researching how mesh credit works, and also how the multilateral offset clearing works. MCS wants, in general to start by offering offset clearing and then have members extend credit to one another to further reduce the need of cash for settlement. So that credit could be supplied either through the mesh which is revealed by the invoices, or by a mesh of trust-lines which reflects personal trust relationships, or by a credit commons like structure which would reflect more group-like credit structures.
> wait, I have those paths backwards, don't I
yes but also the path doesn't contrain a url. Each node in the path is mapped to a url in the account details.
This is a big conversation.

**Reply by JohnG-cc** [reply]
I was just adding the url at the end of the path as a proxy for a trunk name, that would be needed if we were to have a multi-trunk system. But since we aren't, that's not needed.

**Reply by dilgreen** [reply]
I think @matslats is correct in the inference that allowing any node to join multiple trees has an 'economy management' implication as well as a technical /efficiency one.

At the same time, it seems crucial to explore mixing modalities, for several reasons:
- one tree to cover the whole world is clearly ridiculous
- it would be extraordinarily difficult to prevent an entity (individual or business) from joining more than one tree, just as it is hard to stop someone opening accounts with two banks. It would equally be undesirable, would come across as both controlling and paranoid.
    - I generally think it deeply unwise to ignore likely user behaviour one considers either undesirable or too hard to handle.
       - this 'not ignoring' does not, though, necessarily mean 'build technical affordances for'. I think I read that Reddit's approach to complexity is 'let users figure it out for themselves' (I am not necessarily advocating this approach either).

**Reply by dilgreen** [reply]
I have also discussed 'namespaces' - aka identity - aka message provenance -  with both @utunga and @matslats .

I am fairly sure that the long-term defensibility and decentralisation of this network (which I sort of assume will become important long before we are economically significant - bitcoin was attacked on all fronts when the 'market cap' of crypto was negligible compared to what it is now: the finance system is highly alert to potential threats - almost as if it knows it is a house of cards).

And the DNS system is easily used against anything that depends upon it for message provenance.

So I'd like to have thoughts about a namespace system that can be mapped onto DNS (important for now), but is not ultimately dependent upon it (a parallel to the conversation about protocols vs APIs).

**Reply by dilgreen** [reply]
Personally, I would like to explore a ledger that can have more than one BoT account. This reply is about the technical implications.

I don't see this as necessarily too technically complex. Since each BoT a/c has its own balance limit and exchange rate, each is addressed atomically - there is no connection between them. And each user account is atomic in its relations with the BoT accounts, too.

Of course, cc-node would need to be modified - rather than understanding itself as having one 'special' account, it would need to be able to cope with potentially multiple instances of special account type.

Arbitrary complexity could arise, though, through the (highly important) CCP feature which makes it possible for any user to become a node (ie a company could build an internal accounting network as part of an ERP tool). If it were possible for a newly formed leaf to then join a different node on the same tree, strange things might become possible - I can't think it through, but perhaps things like VAT fraud loops or other scams might be achievable. If there was algorithmic credit allocation, then fake trades could make an entity look good.

I think that non-covert versions of this could be avoided, though, by checking for multiple pathways in the same namespace.

Which reminds me that we need a 'white hat sociopath' review of the protocol...

**Reply by dilgreen** [reply]
Further, in discussions with Tomaź Fleishchman, the idea of Mutual Credit Networks having connections in to multilateral offset networks, for the purpose of offering credit in to clearing runs has come up as making a powerful contribution to minimising network debt - decreasing the influence of banks over the economy.

@utunga, as I understand it, also has ideas about connections into the obligation graph to enhance the endogenous clearing power of the network.

**Reply by dilgreen** [reply]
At the same time as all the above, the idea of a strict tree as a mutual, leaf-prioritised, continuous clearing network offers a clarity of mutual empowerment and sovereignty that seems important to preserve.

@tfwoodroofMCS and I have discussed an intuitive feeling that an early proposal I made of a tree with nine tiers allowing a span of dunbar-scale networks all the way from neighbourhoods to a world trunk seems almost certain to be at best unwieldy - and more likely unmanageable / vulnerable to technocratic capture.

I do like @JohnG-cc 's metaphor of a forest.

The relationships between trees in a forest are not achieved by leaves (or roots) being connected to more than one tree (afaik,at least), but through other mechanisms, such as mycelial nets, diffusion of chemicals, and so on.

The key critique I make of mesh networks is that they are;
1. atomising  - since the road to success is to become a superconnector trusted by many (with which the superconnector is unlikely to have a deep relationship).
2. seems to have an inherent need for data aggregators - builders of information asymmetries, which inevitably become power centres.

However, if we imagine Credit Commons trees, where the leaves have privileged access to high quality network data in a structured form, and that these trees have scale which is both sufficient and appropriate (yes, I'm going to use the word bioregion here - while acknowledging that I have no clear definition of a bioregion, or data which proves that these are either sufficient or appropriate), then perhaps connecting trees together with a mesh network is not only acceptable but appropriate? 

At such a scale, the mesh could be mutually owned, and have not-too-many-nodes, and so remain governable. Or perhaps a mesh of meshes?

I am hoping that models like this can be explored before we crystallise too far into a future version of the CCP.

**Reply by tfwoodroofMCS** [reply]
Some follow-ons to @dilgreen's comments, and replying also to @matslats' below:

I've not yet been able to convince myself of the plausibility - at least in the near term - of more than four or five layers of federation. Assuming individual businesses to be the 'first' layer (I haven't spent much time thinking about what internal company ledgers might look like, or how their use might affect the dynamics of 'higher' layers), this gives: 

'Individual businesses' -> 'trading clusters (~20 businesses)' -> 'town/city networks of clusters' -> regional network.

Reasons for being sceptical that any more than this would become unwieldy:

- Although there is some research indicating that real-world trading networks do exhibit scale-free properties (which correspond well to recursive structures), there's clearly a lot more going on, and so a CredCom federation can only expect to immediately & cleanly (i.e. without requiring dramatic changes in existing trading relationships) serve some subset of an economic community. Of course, if the incentives to become part of the federation are strong enough, then over time we may hope to see the convergence @matslats mentions, but I wouldn't want to make any assumptions as to characteristic timescales (or what the unintended consequences of this might be).
- Scale-free properties are an emergent outcome of network dynamics, and these themselves will differ at the neighbourhood, town/city, regional, national, and global scales _as a function of scale_ (entity size & diversity, network density, culture, policy...) in ways that I also wouldn't care to make assumptions about.
- Lack of precedence for the corresponding governance structures: although the nested circles defined by sociocracy seem like a natural fit for CredCom federations, I'd want to find strong evidence that more than a few layers had been sustained over the long term for a variety of cases involving large numbers of otherwise-independent organisations.

Thankfully, I don't think we need a single CredCom tree to be able to cover everything, as we have other tools - multilateral clearing, mesh credit, path resolution - which in combination should support a diversity of scales and structures sufficient to get us most of the way to eliminating the need for bank-money. For me, the question here therefore becomes: 'over what range of scales can we reasonably expect the CredCom Protocol to map onto existing economic structures, and where is the best place to start?' Although I've given my intuitive understanding above, I agree with later comments that the best thing to do is to find out by building a tree, starting - as usual - from the bottom up, and seeing what we learn.

As a final note, Keynes' bancor proposal was for a mutual credit at the global scale, with, as far as I'm aware, nothing to say about organising subsidiary structures in a similar manner. This implies that he considered it (at the time, at least) entirely plausible for governments to manage national trade balances in this way.

**Comment by matslats** [comment]
I don't think that whole world in a single tree is so ridiculous, if convergence benefits all with greater connectivity.
The cc-node (reference implementation) relies quite heavily on their being no more that one trunkward account. - for example if you put in an unknown account name, it will pass it to the trunkwards node to try to parse.
Multiple trunkward nodes implies path finding which implies building a (centralised?) map of the network and keeping it up to date (or querying it in realitime) I just think it's a big leap. 
I do see the value of mesh credit and I agree that we should be thinking about whether and how to deploy it. Mesh credit is better:
- in ungoverned spaces or where there isn't the willingness to do mutual trust. Maybe it would well at the global level where countries have shifting alliances, for example. 
- when there is less desire to balance trade, and participants are happy to settle imbalances with a commodity like dollars or gold.
@dilgreen  note that myCHIPS (Kyle) claims to have a decentralised pathfinding algo.

**Comment by JohnG-cc** [comment]
@matslats it's an interesting point that at a global level there is less trust than local, and that lower trust structures like a mesh may be more at home there... as @dilgreen has said, "trust doesn't scale".

Also I don't understand the need for a path finding algorithm when using the Path Resolution model I outlined - each payee knows their own full path, and communicates this to the payer directly, out of band of CredCom.

**Comment by JohnG-cc** [comment]
Also, I think the main issue is not so much the single tree, but rather the implied single trunk node. 

Out of all possible topologies, surely that would be the second-hardest possible design to secure against censorship, and the second-worst possible scalability and availability challenge, superior only to a monolithic platform and inherently more vulnerable than any design that allows for more than one path between payer and payee?

**Comment by JohnG-cc** [comment]
@dilgreen  wrote:
> Arbitrary complexity could arise, though, through the (highly important) CCP feature which makes it possible for any user to become a node (ie a company could build an internal accounting network as part of an ERP tool). If it were possible for a newly formed leaf to then join a different node on the same tree, strange things might become possible

The idea I outlined above still retained the restriction of a node only being able to have a single BoT account _per tree_, making loops like this impossible. 

Each tree is completely orthogonal to every other tree, and knows nothing about them. Transactions too can still only move inside a single tree. In such a model, it's not the nodes themselves that are leaves or branches, but rather _BoT accounts within a node_ that are leaves and branches of various trees. Natural arboreal ecosystems lack a good metaphor for the role of Node ledger here that groups various BoT and user accounts under a single database - unless perhaps an introduced thing like one of those tents you can suspend off the ground with ropes tied to branches of various trees... This is because the trees are orthogonal dimensions in an n-dimensional space, which maps poorly to our intuitions.

In a more complex cross-tree protocol that allowed routing however, a transaction that could hop across trees like a squirrel is an obvious metaphor...

Having said that though, I am really starting to warm to @matslats observation that mesh credit might be a better trust model at global scales. And linking trees with a mesh _does_ have a nice arboreal metaphor of a spider's web between trees, which is a **lot** easier to communicate than the mind-fuck of extra-dimensional trees overlapping each other in space.

**Comment by utunga** [comment]
@JohnG-cc wrote:
>  I think the main issue is not so much the single tree, but rather the implied single trunk node.

I 100% agree that the concern is not so much the single tree as the single trunk node. A single trunk node is fragile in a few different senses. Firstly because too much ends up depending on the stability of whatever org runs that root node, but also because it creates a single point to 'attack' - this attack could be technical (as with attacks on DNS, say) or it could be organizational (for instance an investor decides to invest in CC on the basis that they can erect a toll gate at that root node and make all their money that way). 

That said, I am starting to wonder if this problem is too hard to address in *this* iteration of our work. 

Even if we assume a somewhat expanded scope for our work here it seems to me that the goal of comprehensively removing that root-node dependency may become a very distracting goal. It's going to be a lot of work just to be really confident that we we can reliably route mutual credit transactions via a single tree. Creating 'multi-tree' routing may be significantly more difficult and distracting. Furthermore I have concerns that conceiving of this as a problem with 'routing' between trees may not even be the best way to think of this problem, that is,  instead of seeing 'multi-tree' routing as being a technical issue instead of, primarily, a governance, counter-party-risk, chaordic organisation type issue. 

So! Perhaps a good compromise would be t rephrase our wording and documentation but, at a technical level, stick to a single tree approach for now. We can then say that our organizational effort is focused on building our first tree or first few, independent, saplings, and that we hope one day to meld them all together in some (as yet unspecified) way. We can stick with the idea of it being recursive and fractal.

This would let us keep our focus on mutual credit and the simpler of use cases for shared payment between trusted organisations and give us a more time to figure out a better approach for the more general approach. 

I dunno I could be convinced either way, but thought I'd raise it as an idea for feedback.

**Reply by JohnG-cc** [reply]
@utunga wrote:
> We can then say that our organizational effort is focused on building our first tree or first few, independent, saplings, and that we hope one day to meld them all together in some (as yet unspecified) way.

I think that would be wise, and productive. 

I was sort of aiming for that by suggesting a layered protocol, so that the CredCom protocol wouldn't need any changes to accommodate it, but of course any software that supports the new protocol _would_ need to implement a new layer to use it. 

I think it does makes sense to defer that effort until we have more than one active tree, and to then make a new layered protocol rather than altering the CredCom protocol, so that individual software implementations can add it while still continuing to interact via the CredCom protocol with other software that has not upgraded?

**Comment by utunga** [comment]
On a very different note I thought I would share some links about Lightning Network (LN) Payment Routing. 

https://docs.lightning.engineering/lightning-network-tools/lnd/pathfinding 

https://medium.com/@renepickhardt/evaluating-path-finding-strategies-on-the-lightning-network-238fe2fdf3d6 

https://lnrouter.app/get-route?sat=9000000&source=033878501f9a4ce97dba9a6bba4e540eca46cb129a322eb98ea1749ed18ab67735&target=036b53093df5a932deac828cca6d663472dbc88322b05eec1d42b26ab9b16caa1c

Obviously the concerns and priorities in LN Routing are very different (being zero trust payment as opposed to high trust credit) but, as is the way with these things, the technical issues and the organizational trust issues will definitely have echoes that might be informative. And of course their work is considerably further advanced than what we have achieved to date.

**Comment by utunga** [comment]
> Having said that though, I am really starting to warm to [@matslats](https://github.com/matslats) observation that mesh credit might be a better trust model at global scales. And linking trees with a mesh does have a nice arboreal metaphor of a spider's web between trees, which is a lot easier to communicate than the mind-fuck of extra-dimensional trees overlapping each other in space.

My thoughts exactly. Though perhaps another metaphor we could use woudl be the famous [mycelial network](https://www.nationalforests.org/blog/underground-mycorrhizal-network) between trees.

**Comment by JohnG-cc** [comment]
@utunga wrote:
> perhaps another metaphor we could use woudl be the famous mycelial network between trees

That's a good metaphor for trunk-to-trunk connections. I guess it depends on what topology we end up with?

**Reply by utunga** [reply]
ha.. true that

**Comment by JohnG-cc** [comment]
@utunga wrote:
> Lightning Network 

A very good (perhaps canonical?) example of a layer 2 payment protocol. 

I have never looked into it deeply, I'll have a read of those!

**Comment by JohnG-cc** [comment]
"LND will prioritize routes that have a good tradeoff between the highest probability of succeeding and the ones that pay the lowest fees"

Sounds like some pretty similar concerns right out of the gate...

**Comment by JohnG-cc** [comment]
@matslats wrote:
> myCHIPS (Kyle) claims to have a decentralised pathfinding algo

I wonder if, when we get to that stage, the myCHIPS guys might be interested in getting involved with designing an inter-tree protocol? Or if myCHIPS might have attributes that could make it a candidate for use in connecting trees? I remember being quite impressed with their model.

But anyway, that's a discussion for a later time, I'm sure.

**Comment by matslats** [comment]
I'm appreciating all your contributions here.
@JohnG-cc you said
< I don't understand the need for a path finding algorithm when using the Path Resolution model I outlined - each payee knows their own full path, and communicates this to the payer directly, out of band of CredCom.
Note that if a node has two trunkward connections then every account on that node and under it has two absolute addresses and therefore two possible paths. A pathfinding algo would want to check
- the cheapest route since each node along the path is likely to apply a transaction fee
- the route that would most restore the balance of trade on the nodes it passes.
The reason you can't 'pay' someone in another tree is that no payment is happening, just recording a debt for later cancellation. Another tree has another unit of debt and no context for promises to be honoured. So the solution is either 
- join trees under another trunkwards node, 
- make another type of reciprical agreement (e.g. trustlines i.e. mesh)
- settle with a commodity, but if that was an option you wouldn't be using a credit system
- settle with money
Interestingly Ripple.com they introduced the idea of a gateway which is in fact a bank's node in the mesh. Since bank credit is money, any trustline backstopped by a bank can be considered a settlement rather than a promise. Banks' presence in the mesh thus enable p2p credit to be converted to fiat money. The credit commons would allow the same possibility

**Comment by JohnG-cc** [comment]
Right, I get what you mean about pathfinding now - I had envisioned that to occur in the L2 protocol, which would consider which of the trees that payer and payee were mutually part of would lead to lower fees etc. - and then it would pick a tree and the transaction would occur entirely within that tree, with no modification to the protocol required therefore. 

In this model you still depend on both payer and payee being in the same tree, it doesn't really aggregate trees together, it just leverages the fact that nodes could conceivably sign up to a whole portfolio of trees, and then use any of them by switching the trunkwards account pointer to different accounts.

I can still see that potentially being a value add for some users in future, but from previous discussion I realise this doesn't provide inter-tree connections, and so doesn't help build a global distributed network of trees linked by some L2 protocol.

**Comment by JohnG-cc** [comment]
I thought @utunga 's suggestion, of leaving inter-tree connections as a to-do for some yet to be determined level 2 protocol so that we can focus for now on productionising single trees, sounded like a good idea? My idea of facilitating multi-membership by nodes would be ancillary to that, and not something to worry about for now.

**Reply by dilgreen** [reply]
This makes sense to me, too. MCS certainly needs to build a single tree, to show what CredCom can do - and to make good on its business-model proposition and thus justify investment.

I share the general feeling that graphs offer an interesting approach to inter-tree resolution - allowing both for loop clearing and trustline-style settlement between trunks. Allowing for both trustless and trustful modes seems worth thinking about.

**Comment by matslats** [comment]
Allowing people to have accounts on different nodes doesn't break anything, and probably can't be prevented.
However at the governance level it needs to be noted that credit granted to an entity is not necessarily exclusive.
If entities can join multiple nodes, they could access total credit multiple times what they are expected to repay. That suggests that credit should be granted on the basis of trusting account holders at their word, rather than on the basis of ability to pay.

**Reply by dilgreen** [reply]
This post specific to: credit allocation risk in a multi-tree CredCom world. 

Currently @tfwoodroofMCS and I are imagining/intuiting that MC groups will actually be quite small - that the pressures to grow membership numbers evident in our experience in Brixton, with Sardex, or indeed with Bartercard all relate directly to the 'walled garden' characteristic of those networks, and indirectly to the 'build it and they will come' / random model of membership selection.

The model of using clearing networks as a simpler, lower risk proposition for B2B engagement, and high-street shopping vouchers as the equivalent 'gateway drug' for B2C engagement, has a useful outcome in the form of producing near realtime economy maps.
![Graph-060623](https://github.com/CredComSoc/ProtocolWorkingGroup/assets/5934679/2bc66c0b-f72b-45a2-9f8b-0ac2a437bbf3)

Such maps make it possible for us to suggest mutual credit groupings that are well evidenced, the performance of which can be expected to be beneficial to all members - and also support targeted recruiting.

We expect that such identifiable clusters will be relatively small.

Since their internal inter-trade will have an acceptable degree of balance across time, while their connection into a local CredCom federation, and also to the clearing club, will give opportunities to access  a diverse range of trading partners and also to further minimise inevitable settlement needs of their 'home' network, they should experience low pressure to grow other than organically. 

Thus, intuitively, we have the feeling that these groups will stay small.

What this scaling intuition suggests is that credit allocation can be potentially quite social , on the basis of relatively intimate trading relations, and thus understanding of each other's productive capacity - again, on an almost real-time basis. (_we have this fun idea of quarterly or annual 'horse trading' - or 'co-budgeting for credit' - with each member stating their appetite for risk  -maximum exposure aka +ve balance - and desire for credit - maximum negative balance - in a series of rounds, until a balance is reached_).

If a member of a small MC group like this is also a member of another MC group, this will likely be known. A default term of a Member Agreement would likely be full disclosure of membership of other groups.

If all this is true, then taking on too much credit could largely be a socially controllable risk.
